Updated CookieConsent to version 3.0.0. and some minor updates

rahkapetteri opened the pull request:
Hi!

I think Cookie Consent 3.0.0 is working properly now 👨‍💻

Additional changes:

- Added the option to declare cookies (even with regex variables) so cookies are getting deleted after changing consent
- Fixed issues with embedded content when changing consent. Before the fix, when consent was removed, I could still see the embeds
- Updated documentation
- Updated Iframe Manager to 1.2.5
- Added cc_link CSS-style to assets/cookieconsent.css
- Fixes #2 
 
If you have any questions about the fixes I’m happy to discuss them 😊

rahkapetteri committed:
WIP: Migrating to CookieConsent 3. Fix consent data format sent to backend, function/someway to add cookie table, check config file + test everything

rahkapetteri committed:
WIP: Migrating to CookieConsent 3. Fixed: embed consent state, data sending to backend and custom attributes. Updated iFramemanager to 1.2.5.

rahkapetteri committed:
Updated readme, added option to declare cookies, iframemanager consent change update

rahkapetteri committed:
Fixes for embeds consent changing, added google consent mode for analytics_storage if google cookies defined

rahkapetteri committed:
Something weird happened

rahkapetteri committed:
Something weird happened

rahkapetteri committed:
Documentation fix

rahkapetteri committed:
Removed GA detection script

rahkapetteri committed:
Changed default style

rahkapetteri committed:
Updated defaults, readme and added cc-link style from 2.9.

rahkapetteri committed:
pref modal default change

rahkapetteri committed:
Modal width fix + button default change

rahkapetteri committed:
Removed button from string

rahkapetteri committed:
fix indentation

rahkapetteri committed:
fix indentation

rahkapetteri committed:
Deleted empty file

rahkapetteri committed:
fix

rahkapetteri committed:
expire date fix + indendation

rahkapetteri committed:
-data-cookiecategory replaces

rahkapetteri committed:
Iframe Manager required to be run every page load to detect iframe correctly

rahkapetteri committed:
Removed miscopy

rahkapetteri committed:
removed not needed cookie removal

rahkapetteri committed:
Disable autoupdater and add latest change from source branch

rahkapetteri committed:
Added forgotten edited info

raikasdev commented on air-cookie.php:
```diff
@@ -57,12 +59,12 @@ function get_script_version() {
 /**
- * Github updater.
+ * Github updater. Disabled by commenting out.
  *
  * @since 0.1.0
  */
-require plugin_base_path() . '/plugin-update-checker/plugin-update-checker.php';
-$update_checker = \Puc_v4_Factory::buildUpdateChecker( 'https://github.com/digitoimistodude/air-cookie', __FILE__, 'air-cookie' );
+// require plugin_base_path() . '/plugin-update-checker/plugin-update-checker.php';
+// $update_checker = \Puc_v4_Factory::buildUpdateChecker( 'https://github.com/digitoimistodude/air-cookie', __FILE__, 'air-cookie' );
```
These kinds of changes shouldn't be included in PRs :smiley:

rahkapetteri replied:
Was meant for local only 🤦‍♂️

raikasdev replied:
Haha, happens to the best of us.

raikasdev commented on settings.php:
```diff
@@ -26,21 +26,35 @@ function get_settings() {
-    'gui_options'       => [
-      'consent_modal' => [
-        'layout'    => 'box',
-        'position'  => 'bottom left',
+    'revision'              => $categories_version, // use version number to invalidate if categories change
+    'cookie'                => [
+      'name' => 'air_cookie',
+      'expiresAfterDays' => 182, // in days, 182 days = 6 months
+    ],
+
+    'guiOptions'     => [
+      'consentModal' => [
+        'layout'     => 'box wide',
+        'position'   => 'bottom left',
+        'equalWeightButtons' => false,
+        'flipButtons'        => false,
+      ],
+
+      'preferencesModal' => [
+        'layout'             => 'box',
+        'equalWeightButtons' => true,
+        'flipButtons'        => false,
       ],
     ],
+
+    'language'  => [
+      'default' => $lang,
+      'translations'   => [
+        $lang   => []
+      ]
+    ],
+
+    'categories' => [],
   ];
```
Refactoring the settings object isn't a good practice as this will break backwards compability pretty much just for style points.

We name variables according to the WordPress Coding Standard's [Naming Conventions](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#naming-conventions) meaning `snake_case` instead of `camelCase`.

raikasdev replied:
Changes are from CookieConsent, ahh.

raikasdev replied:
I'd still prefer at least mapping the old keys to the new ones if they are set :thinking:

rahkapetteri replied:
I´ll look into it.

raikasdev commented on README.md:
````diff
@@ -60,13 +61,68 @@ function my_add_cookie_category( $categories ) {
 
 There is also `air_cookie\categories\{category-key}` filter available to change the settings of indivual category.
+```php
+add_filter( 'air_cookie\categories\{analytics}', 'my_change_category_analytics' );
````
It would be `air_cookie\categories\analytics`, without the braces.

rahkapetteri replied:
Hmm I can´t get it to work without brackets. Only way to use the filter without brackets would require to change in settings.php:

`$categories[ $key ] = apply_filters( "air_cookie\categories\{$category_key}", $category );`

to `$categories[ $key ] = apply_filters( "air_cookie\categories\\{$category_key}", $category );`

Atleast that´s how [PHP-docs say](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc) to handle backslashed variables 🤔

`// Won't work, outputs: C:\folder\{fantastic}.txt`
`echo "C:\folder\{$great}.txt"`

`// Works, outputs: C:\folder\fantastic.txt`
`echo "C:\\folder\\{$great}.txt"`

raikasdev replied:
Not in the code, I meant in the example :smile: 

If user wants to filter analytics they need to create a filter `add_filter ( 'air_cookie\categories\analytics', 'my_change_analytics' );`, which is then called by `apply_filters( "air_cookie\categories\{$category_key}", $category );`.

`"air_cookie\categories\{$category_key}"` gets resolved to `"air_cookie\categories\analytics` as `{$category_key}` is a string template.

raikasdev replied:
This is indeed a bug. #10

raikasdev commented on script-injection.php:
```diff
@@ -33,67 +33,125 @@ function inject_js() {
   $settings = get_settings(); // phpcs:ignore WordPress.WP.DeprecatedFunctions.get_settingsFound
   if ( ! is_array( $settings ) ) {
-    return;
+		return;
   }
 
   // Cookie Consent javascript base.
-  wp_enqueue_script( 'cookieconsent', plugin_base_url() . '/assets/cookieconsent.js', [], get_script_version(), false );
+  wp_enqueue_script( 'cookieconsent', plugin_base_url() . '/assets/cookieconsent.js', [], get_script_version(),
+  array(
+    'in_footer' => true,
+    'strategy'  => 'defer',
+  )
+);
```
What is happening here with the whitespace?

rahkapetteri replied:
Probably VS Code auto indentation with different settings 😅

raikasdev commented on script-injection.php:
```diff
@@ -33,67 +33,125 @@ function inject_js() {
   // Cookie Consent javascript base.
-  wp_enqueue_script( 'cookieconsent', plugin_base_url() . '/assets/cookieconsent.js', [], get_script_version(), false );
+  wp_enqueue_script( 'cookieconsent', plugin_base_url() . '/assets/cookieconsent.js', [], get_script_version(),
+  array(
+    'in_footer' => true,
+    'strategy'  => 'defer',
+  )
+);
```
We use the short [] syntax instead of array().

raikasdev commented on script-injection.php:
```diff
@@ -33,67 +33,125 @@ function inject_js() {
 
+      <?php // Add functions to handle changes ?>
+      const ccOnChanges = {
+        onFirstConsent: () => {
+          ccOnAccept();
+        },
+
+        onConsent: () => {
+          checkIframeConsent();
+        },
+
+        onModalShow: () => {
+          <?php if ( apply_filters( 'air_cookie\styles\set_max_width', true ) ) : ?>
+            var cookieconsent_element = document.querySelector('div#cc-main div.cm');
+            if( typeof( cookieconsent_element ) != 'undefined' && cookieconsent_element != null ) {
+              cookieconsent_element.style = 'max-width: 40em;';
+            }
+          <?php endif; ?>
+        },
+
+        onChange: () => {
+          ccOnChange();
+        }
```
onFirstConsent, onConsent and onChange probably don't need to be inside a arrow function and can be passed as arguments.

`onChange: ccOnChange`.

raikasdev commented on script-injection.php:
```diff
@@ -145,8 +204,7 @@ function do_category_js( $category ) {
   ob_start(); ?>
-
-  if ( cc.allowedCategory( '<?php echo $category_key; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>' ) ) {
+  if ( CookieConsent.getCookie( 'categories' ).includes( '<?php echo $category_key; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>' ) ) {
```
Why do we have to use this instead of [`CookieConsent.acceptedCategory()`](https://cookieconsent.orestbida.com/reference/api-reference.html#acceptedcategory)?

rahkapetteri replied:
Wasn't working correctly in 3.0.0, but I think that´s fixed in 3.0.1 now (PR is 3.0.0).

raikasdev replied:
Ah!

raikasdev commented on rest-api.php:
```diff
@@ -48,9 +50,16 @@ function record_consent( $request ) {
-  // Try if the user consent for this revision and levels has been already recorded
-  // Test query: SELECT id FROM wp_air_cookie WHERE visitor_id = 'f284009a-ace9-42e7-a5ef-42b065a9184c' AND cookie_revision = '2412150750' AND cookie_value = 'a:4:{i:0;s:9:"necessary";i:1;s:10:"functional";i:2;s:9:"analytics";i:3;s:6:"embeds";}'
-  $exists = $wpdb->get_row( $wpdb->prepare( "SELECT id FROM {$table_name} WHERE visitor_id = %s AND cookie_revision = %s AND cookie_value = %s", $data->visitorid, $data->revision, $cookie_value ) ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.PreparedSQLPlaceholders.QuotedSimplePlaceholder, WordPress.DB.PreparedSQL.Interpo ...
+  // Check if the index exists
+  $index_exists = $wpdb->get_row( "SHOW INDEX FROM {$table_name} WHERE Key_name = 'idx_id_revision_value'" ); // phpcs:ignore
+
+  // CREATE INDEX idx_id_revision_value ON wp_air_cookie (visitor_id, cookie_revision, cookie_value);
+  if ( null === $index_exists ) {
+    $wpdb->query( "CREATE INDEX idx_id_revision_value ON {$table_name} (visitor_id, cookie_revision, cookie_value);" ); // phpcs:ignore
+  }
```
Does this have to happen on every request or could this just be ran when the plugin is activated or if the version has changed?

rahkapetteri commented:
Thanks for comments! I´ll look into those.

I was doing some testing with fork repo and didn´t notice latest changes went to PR also.

rollecode commented:
> Thanks for comments! I´ll look into those.
> 
> I was doing some testing with fork repo and didn´t notice latest changes went to PR also.

What's the status? We are adding this to our roadmap as well, but if you are ahead of us, it all helps. Thanks.

rahkapetteri committed:
Added CookieConsent 3.0.1, snake cased settings, fixed filter names and minor formatting

rahkapetteri commented:
> > Thanks for comments! I´ll look into those.
> > I was doing some testing with fork repo and didn´t notice latest changes went to PR also.
> 
> What's the status? We are adding this to our roadmap as well, but if you are ahead of us, it all helps. Thanks.

Hello, I pushed latest changes which I have done. Added latest CookieConsent version, fixed https://github.com/digitoimistodude/air-cookie/issues/10#issue-2401183270, fixed some minor formatting and created function to allow old settings formatting, so old installations shouldn't break (but all old used settings should still be checked).

Hope these help.

rahkapetteri committed:
3.0.1 js and enabled autoupdated
